File: ios/OrderedListMarkerFormatter.swift
import Foundation

enum EditorOrderedListNumberingScheme: String {
    case decimal
    case lowerAlpha
    case upperAlpha
    case lowerRoman
    case upperRoman
}

struct EditorOrderedListMarkerTheme {
    static let defaultSchemes: [EditorOrderedListNumberingScheme] = [
        .decimal,
        .lowerAlpha,
        .lowerRoman
    ]

    let schemes: [EditorOrderedListNumberingScheme]
    let suffix: String

    init(dictionary: [String: Any]) {
        let parsed = (dictionary["schemes"] as? [Any])?.compactMap { value in
            (value as? String).flatMap(EditorOrderedListNumberingScheme.init(rawValue:))
        } ?? []
        schemes = parsed.isEmpty ? Self.defaultSchemes : parsed
        let parsedSuffix = dictionary["suffix"] as? String
        suffix = parsedSuffix == ")" ? ")" : "."
    }
}

enum OrderedListMarkerFormatter {
    static func label(
        index: UInt32,
        nestingDepth: Int,
        theme: EditorOrderedListMarkerTheme?
    ) -> String {
        let resolved = theme ?? EditorOrderedListMarkerTheme(dictionary: [:])
        let depth = max(0, nestingDepth)
        let scheme = resolved.schemes[depth % resolved.schemes.count]
        return formattedIndex(index, scheme: scheme) + resolved.suffix
    }

    private static func formattedIndex(
        _ index: UInt32,
        scheme: EditorOrderedListNumberingScheme
    ) -> String {
        switch scheme {
        case .decimal:
            return String(index)
        case .lowerAlpha:
            return alphabeticIndex(index) ?? String(index)
        case .upperAlpha:
            return alphabeticIndex(index)?.uppercased() ?? String(index)
        case .lowerRoman:
            return romanIndex(index) ?? String(index)
        case .upperRoman:
            return romanIndex(index)?.uppercased() ?? String(index)
        }
    }

    private static func alphabeticIndex(_ index: UInt32) -> String? {
        guard index > 0 else { return nil }

        var value = index
        var characters: [String] = []
        while value > 0 {
            let offset = (value - 1) % 26
            guard let scalar = UnicodeScalar(97 + offset) else { return nil }
            characters.append(String(scalar))
            value = (value - 1) / 26
        }
        return characters.reversed().joined()
    }

    private static func romanIndex(_ index: UInt32) -> String? {
        guard index > 0, index <= 3_999 else { return nil }

        let table: [(value: UInt32, symbol: String)] = [
            (1_000, "m"),
            (900, "cm"),
            (500, "d"),
            (400, "cd"),
            (100, "c"),
            (90, "xc"),
            (50, "l"),
            (40, "xl"),
            (10, "x"),
            (9, "ix"),
            (5, "v"),
            (4, "iv"),
            (1, "i")
        ]
        var value = index
        var result = ""
        for entry in table {
            while value >= entry.value {
                result += entry.symbol
                value -= entry.value
            }
        }
        return result
    }
}

File: android/src/main/java/com/apollohg/editor/OrderedListMarkerFormatter.kt
package com.apollohg.editor

import org.json.JSONObject

enum class EditorOrderedListNumberingScheme {
    DECIMAL,
    LOWER_ALPHA,
    UPPER_ALPHA,
    LOWER_ROMAN,
    UPPER_ROMAN;

    companion object {
        fun fromJson(value: String): EditorOrderedListNumberingScheme? = when (value) {
            "decimal" -> DECIMAL
            "lowerAlpha" -> LOWER_ALPHA
            "upperAlpha" -> UPPER_ALPHA
            "lowerRoman" -> LOWER_ROMAN
            "upperRoman" -> UPPER_ROMAN
            else -> null
        }
    }
}

private val DEFAULT_ORDERED_LIST_SCHEMES = listOf(
    EditorOrderedListNumberingScheme.DECIMAL,
    EditorOrderedListNumberingScheme.LOWER_ALPHA,
    EditorOrderedListNumberingScheme.LOWER_ROMAN
)

data class EditorOrderedListMarkerTheme(
    val schemes: List<EditorOrderedListNumberingScheme> = DEFAULT_ORDERED_LIST_SCHEMES,
    val suffix: String = "."
) {
    companion object {
        fun fromJson(json: JSONObject?): EditorOrderedListMarkerTheme? {
            json ?: return null
            val parsedSchemes = json.optJSONArray("schemes")?.let { values ->
                buildList {
                    for (index in 0 until values.length()) {
                        val rawValue = values.opt(index) as? String ?: continue
                        EditorOrderedListNumberingScheme.fromJson(rawValue)?.let(::add)
                    }
                }
            }.orEmpty()

            return EditorOrderedListMarkerTheme(
                schemes = parsedSchemes.ifEmpty {
                    DEFAULT_ORDERED_LIST_SCHEMES
                },
                suffix = if (json.optString("suffix") == ")") ")" else "."
            )
        }
    }
}

internal object OrderedListMarkerFormatter {
    private const val MAX_INDEX = 4_294_967_295L

    private val romanTable = listOf(
        1_000L to "m",
        900L to "cm",
        500L to "d",
        400L to "cd",
        100L to "c",
        90L to "xc",
        50L to "l",
        40L to "xl",
        10L to "x",
        9L to "ix",
        5L to "v",
        4L to "iv",
        1L to "i"
    )

    fun label(index: Long, nestingDepth: Int, theme: EditorOrderedListMarkerTheme?): String {
        val resolvedTheme = theme ?: EditorOrderedListMarkerTheme()
        val schemes = resolvedTheme.schemes.ifEmpty {
            DEFAULT_ORDERED_LIST_SCHEMES
        }
        val scheme = schemes[nestingDepth.coerceAtLeast(0) % schemes.size]
        val suffix = if (resolvedTheme.suffix == ")") ")" else "."
        return formattedIndex(index, scheme) + suffix
    }

    private fun formattedIndex(index: Long, scheme: EditorOrderedListNumberingScheme): String {
        if (index !in 0..MAX_INDEX) return index.toString()

        return when (scheme) {
            EditorOrderedListNumberingScheme.DECIMAL -> index.toString()

            EditorOrderedListNumberingScheme.LOWER_ALPHA -> alphabeticIndex(index)
                ?: index.toString()

            EditorOrderedListNumberingScheme.UPPER_ALPHA ->
                alphabeticIndex(index)?.uppercase() ?: index.toString()

            EditorOrderedListNumberingScheme.LOWER_ROMAN -> romanIndex(index) ?: index.toString()

            EditorOrderedListNumberingScheme.UPPER_ROMAN ->
                romanIndex(index)?.uppercase() ?: index.toString()
        }
    }

    private fun alphabeticIndex(index: Long): String? {
        if (index == 0L) return null

        var value = index
        val result = StringBuilder()
        while (value > 0) {
            val offset = ((value - 1) % 26).toInt()
            result.append(('a'.code + offset).toChar())
            value = (value - 1) / 26
        }
        return result.reverse().toString()
    }

    private fun romanIndex(index: Long): String? {
        if (index !in 1..3_999) return null

        var value = index
        return buildString {
            for ((unit, symbol) in romanTable) {
                while (value >= unit) {
                    append(symbol)
                    value -= unit
                }
            }
        }
    }
}
